File: tebd/tebd_pott_chain.py
import numpy as np
from scipy.linalg import expm
from numpy.linalg import svd

###############################################
# Local Potts operators
###############################################

def potts_operators():
    w = np.exp(2j*np.pi/3)

    Z = np.array([
        [1, 0, 0],
        [0, w, 0],
        [0, 0, w**2]
    ], dtype=complex)

    X = np.array([
        [0, 0, 1],
        [1, 0, 0],
        [0, 1, 0]
    ], dtype=complex)

    I = np.eye(3, dtype=complex)
    return Z, X, I


###############################################
# Bond Hamiltonians with correct onsite splitting
###############################################

def potts_H_bond(J, f, L):
    Z, X, I = potts_operators()
    Xh = X.conj().T

    H_bond = []

    for i in range(L - 1):
        H = np.zeros((9, 9), dtype=complex)

        # nearest-neighbor Potts interaction
        H += -J[i] * (np.kron(Z.conj().T, Z) + np.kron(Z, Z.conj().T))

        # split transverse field between neighboring bonds
        left_weight  = 1.0 if i == 0     else 0.5
        right_weight = 1.0 if i == L - 2 else 0.5

        H += -left_weight  * f[i]   * (np.kron(X, I) + np.kron(Xh, I))
        H += -right_weight * f[i+1] * (np.kron(I, X) + np.kron(I, Xh))

        H_bond.append(H.reshape(3, 3, 3, 3))

    return H_bond


def potts_expH_bond(J, f, L, delta):
    H_bond = potts_H_bond(J, f, L)
    expH_bond = []

    for H in H_bond:
        U = expm(-delta * H.reshape(9, 9))
        expH_bond.append(U.reshape(3, 3, 3, 3))

    return expH_bond, H_bond


###############################################
# Initial MPS in Vidal form
###############################################

def initialMat(L, d, chiMax):
    G = []
    l = []
    chi = [0] * (L + 1)

    for pos in range(0, L // 2 + 1):
        chi[pos] = min(chiMax, d**pos)
    for pos in range(L // 2 + 1, L + 1):
        chi[pos] = min(chiMax, d**(L - pos))

    for pos in range(L):
        chi0 = int(chi[pos])
        chi1 = int(chi[pos + 1])
        G.append(np.zeros((d, chi0, chi1), dtype=np.complex128))
        l.append(np.ones(chi0, dtype=np.complex128))

    l.append(np.ones(int(chi[L]), dtype=np.complex128))

    # product state |0...0>
    for pos in range(L):
        G[pos][0, 0, 0] = 1.0

    return l, G


###############################################
# One TEBD sweep on chosen bond parity
###############################################

def tebd_sweep_parity(l, G, expH, parity):
    L = len(G)

    for i_bond in range(parity, L - 1, 2):
        i0 = i_bond
        i1 = i_bond + 1
        i2 = i_bond + 2

        d = G[i0].shape[0]
        chi0 = G[i0].shape[1]
        chi1 = G[i0].shape[2]
        chi2 = G[i1].shape[2]

        l0inv = 1.0 / (l[i0] + 1e-20)
        l2inv = 1.0 / (l[i2] + 1e-20)

        # theta tensor
        Psi = np.tensordot(np.diag(l[i0]), G[i0], axes=(1, 1))
        Psi = np.tensordot(Psi, np.diag(l[i1]), axes=(2, 0))
        Psi = np.tensordot(Psi, G[i1], axes=(2, 1))
        Psi = np.tensordot(Psi, np.diag(l[i2]), axes=(3, 0))
        # shape: (chi0, d, d, chi2)

        # apply two-site gate
        Phi = np.tensordot(Psi, expH[i0], axes=([1, 2], [0, 1]))
        Phi = np.transpose(Phi, (0, 2, 3, 1))
        Phi = np.reshape(Phi, (chi0 * d, d * chi2))

        # split with SVD
        U, S, Vh = svd(Phi, full_matrices=False)

        keep = min(chi1, len(S))
        U = U[:, :keep]
        S = S[:keep]
        Vh = Vh[:keep, :]

        # normalize Schmidt values
        nrm = np.linalg.norm(S)
        if nrm > 0:
            S = S / nrm

        l[i1] = S

        # update Gamma[i0]
        U = U.reshape(chi0, d, keep)
        U = np.tensordot(np.diag(l0inv), U, axes=(1, 0))
        U = np.transpose(U, (1, 0, 2))
        G[i0] = U

        # update Gamma[i1]
        V = Vh.reshape(keep, d, chi2)
        V = np.transpose(V, (1, 0, 2))
        V = np.tensordot(V, np.diag(l2inv), axes=(2, 0))
        G[i1] = V


###############################################
# Second-order Suzuki-Trotter
###############################################

def tebd_second_order_step(l, G, expH_half, expH_full):
    tebd_sweep_parity(l, G, expH_half, parity=0)
    tebd_sweep_parity(l, G, expH_full, parity=1)
    tebd_sweep_parity(l, G, expH_half, parity=0)


###############################################
# Energy measurement
###############################################

def bond_energy(l, G, H, pos):
    Psi = np.tensordot(np.diag(l[pos]), G[pos], axes=(1, 1))
    Psi = np.tensordot(Psi, np.diag(l[pos + 1]), axes=(2, 0))
    Psi = np.tensordot(Psi, G[pos + 1], axes=(2, 1))
    Psi = np.tensordot(Psi, np.diag(l[pos + 2]), axes=(3, 0))

    rho = np.tensordot(np.conj(Psi), Psi, axes=([0, 3], [0, 3]))
    return np.tensordot(rho, H[pos], axes=([0, 1, 2, 3], [0, 1, 2, 3]))


def total_energy(l, G, H_bond):
    E = 0.0 + 0.0j
    for i in range(len(G) - 1):
        E += bond_energy(l, G, H_bond, i)
    return E


###############################################
# Main
###############################################

if __name__ == "__main__":
    L = 7
    d = 3
    chiMax = 30
    delta = 0.01
    steps = 1000

    J = np.ones(L, dtype=float)
    f = np.ones(L, dtype=float)

    l, G = initialMat(L, d, chiMax)

    expH_half, H_bond = potts_expH_bond(J, f, L, delta / 2.0)
    expH_full, _ = potts_expH_bond(J, f, L, delta)
    
    print(f"L = {L}")
    for step in range(steps):
        tebd_second_order_step(l, G, expH_half, expH_full)


        if step % 100 == 0:
            E = total_energy(l, G, H_bond)
            print(f"step {step:3d} energy {E.real:.12f}")

    # Final energy measurement
    print("\nFinal energy: ", total_energy(l, G, H_bond)/L)
    print("Exact ground-state energy = -2.355433747522252")
    print("Difference from exact: ", total_energy(l, G, H_bond)/L - (-2.355433747522252))

File: tebd/tebd_pott_chain.jl
using LinearAlgebra
using TensorOperations

###############################################
# Helpers: NumPy-style matrix <-> tensor map
###############################################

function mat_to_tensor4(M::AbstractMatrix{<:Complex})
    return permutedims(reshape(M, 3, 3, 3, 3), (2, 1, 4, 3))
end

function tensor4_to_mat(T::AbstractArray{<:Complex,4})
    return reshape(permutedims(T, (2, 1, 4, 3)), 9, 9)
end

###############################################
# Local Potts operators
###############################################

function potts_operators()
    w = exp(2im * π / 3)

    Z = ComplexF64[
        1 0 0
        0 w 0
        0 0 w^2
    ]

    X = ComplexF64[
        0 0 1
        1 0 0
        0 1 0
    ]

    I3 = Matrix{ComplexF64}(I, 3, 3)
    return Z, X, I3
end

###############################################
# Bond Hamiltonians
###############################################

function potts_H_bond(J, f, L)
    Z, X, I3 = potts_operators()
    Zh = adjoint(Z)
    Xh = adjoint(X)

    H_bond = Vector{Array{ComplexF64,4}}(undef, L - 1)

    for i in 1:(L - 1)
        Hmat = zeros(ComplexF64, 9, 9)

        Hmat .+= -J[i] .* (kron(Zh, Z) + kron(Z, Zh))

        lw = (i == 1)     ? 1.0 : 0.5
        rw = (i == L - 1) ? 1.0 : 0.5

        Hmat .+= -lw * f[i]   .* (kron(X, I3) + kron(Xh, I3))
        Hmat .+= -rw * f[i+1] .* (kron(I3, X) + kron(I3, Xh))

        H_bond[i] = mat_to_tensor4(Hmat)
    end

    return H_bond
end

function potts_expH_bond(J, f, L, δ)
    H_bond = potts_H_bond(J, f, L)
    expH_bond = Vector{Array{ComplexF64,4}}(undef, L - 1)

    for i in 1:(L - 1)
        Hmat = tensor4_to_mat(H_bond[i])
        Umat = exp(-δ * Hmat)
        expH_bond[i] = mat_to_tensor4(Umat)
    end

    return expH_bond, H_bond
end

###############################################
# Initial MPS in Vidal form
###############################################

function initial_mat(L, d)
    G = [zeros(ComplexF64, d, 1, 1) for _ in 1:L]
    l = [ones(Float64, 1) for _ in 1:(L + 1)]

    for i in 1:L
        G[i][1, 1, 1] = 1.0 + 0im
    end

    return l, G
end

###############################################
# Utilities
###############################################

function safe_inv(v; cutoff=1e-14)
    out = zeros(Float64, length(v))
    for i in eachindex(v)
        if abs(v[i]) > cutoff
            out[i] = 1.0 / v[i]
        end
    end
    return out
end

function left_scale_gamma(lvec, G)
    d, χL, χR = size(G)
    out = similar(G)
    for a in 1:χL
        out[:, a, :] .= lvec[a] .* G[:, a, :]
    end
    return out
end

function right_scale_gamma(G, lvec)
    d, χL, χR = size(G)
    out = similar(G)
    for b in 1:χR
        out[:, :, b] .= G[:, :, b] .* lvec[b]
    end
    return out
end

###############################################
# One TEBD sweep
###############################################

function tebd_sweep_parity!(l, G, expH, parity, chiMax)
    L = length(G)
    start_idx = (parity == 0) ? 1 : 2

    for i in start_idx:2:(L - 1)
        χL = size(G[i], 2)
        χM = size(G[i], 3)
        χR = size(G[i+1], 3)
        d = size(G[i], 1)

        invL = safe_inv(l[i])
        invR = safe_inv(l[i+2])

        # Build theta[a,s,t,b] in steps
        A = left_scale_gamma(l[i], G[i])         # (s,a,c)
        B = right_scale_gamma(A, l[i+1])         # (s,a,c)

        @tensor C[s,t,a,b] := B[s,a,c] * G[i+1][t,c,b]

        theta = Array{ComplexF64}(undef, χL, d, d, χR)
        for b in 1:χR
            theta[:, :, :, b] .= permutedims(C[:, :, :, b], (3, 1, 2)) .* l[i+2][b]
        end
        # theta[a,s,t,b]

        # Apply gate
        @tensor phi[a,s,t,b] := theta[a,x,y,b] * expH[i][x,y,s,t]

        # Reshape for SVD: rows=(a,s), cols=(t,b)
        phi_mat = reshape(phi, χL * d, d * χR)

        F = svd(phi_mat)
        keep = min(length(F.S), chiMax)

        U  = F.U[:, 1:keep]
        S  = F.S[1:keep]
        Vt = F.Vt[1:keep, :]

        nrm = norm(S)
        if nrm > 0
            S ./= nrm
        end
        l[i+1] = S

        # Left Gamma update
        Uten = reshape(U, χL, d, keep)
        Uten = permutedims(Uten, (2, 1, 3))  # (s,a,α)
        for a in 1:χL
            Uten[:, a, :] .*= invL[a]
        end
        G[i] = ComplexF64.(Uten)

        # Right Gamma update
        Vten = reshape(Vt, keep, d, χR)      # (α,t,b)
        Vten = permutedims(Vten, (2, 1, 3))  # (t,α,b)
        for b in 1:χR
            Vten[:, :, b] .*= invR[b]
        end
        G[i+1] = ComplexF64.(Vten)
    end

    return nothing
end

###############################################
# Energy measurement
###############################################

function bond_energy(l, G, H_bond, i)
    A = left_scale_gamma(l[i], G[i])
    B = right_scale_gamma(A, l[i+1])

    @tensor C[s,t,a,b] := B[s,a,c] * G[i+1][t,c,b]

    χR = size(C, 4)
    χL = size(C, 3)
    d1 = size(C, 1)
    d2 = size(C, 2)

    theta = Array{ComplexF64}(undef, χL, d1, d2, χR)
    for b in 1:χR
        theta[:, :, :, b] .= permutedims(C[:, :, :, b], (3, 1, 2)) .* l[i+2][b]
    end

    @tensor rho[s,t,s2,t2] := conj(theta[a,s,t,b]) * theta[a,s2,t2,b]

    return sum(rho .* H_bond[i])
end

function energy_measurement(l, G, H_bond)
    E_total = 0.0
    for i in 1:(length(G) - 1)
        E_total += real(bond_energy(l, G, H_bond, i))
    end
    return E_total
end

###############################################
# Main
###############################################

function main()
    L = 7
    d = 3
    chiMax = 30
    delta = 0.01
    steps = 1000

    J = ones(Float64, L)
    f = ones(Float64, L)

    l, G = initial_mat(L, d)
    expH_half, H_list = potts_expH_bond(J, f, L, delta / 2)
    expH_full, _      = potts_expH_bond(J, f, L, delta)

    println("Starting TEBD for L=$L Potts chain...")

    for s in 1:steps
        tebd_sweep_parity!(l, G, expH_half, 0, chiMax)
        tebd_sweep_parity!(l, G, expH_full, 1, chiMax)
        tebd_sweep_parity!(l, G, expH_half, 0, chiMax)

        if s % 100 == 0
            E = energy_measurement(l, G, H_list)
            println("Step $s | Energy: $(round(E, digits=12))")
        end
    end

    final_E = energy_measurement(l, G, H_list)
    println("\nFinal Energy: $final_E/L = $(final_E/L)")
    println("Exact Energy /L: -2.355433747522252")
    println("Difference:   $(final_E/L - (-2.355433747522252))")
end

main()
